From fc2282d1d53d60d6ee72f3991e464c7d6310bea0 Mon Sep 17 00:00:00 2001
From: Jean Christophe Beyler <jean.christophe.beyler@intel.com>
Date: Tue, 23 Apr 2013 14:26:46 -0700
Subject: Dalvik: Preheader Block Should not be Included in the Loop

BZ: 103339

Removing the preheader block from the LoopInformation, it is not part of the
loop

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I6a9d48eba413eecfc6ffbbec5217a60a548f2898
Orig-MCG-Change-Id: Iff01d98890101e59d44e66c09cfd86b5d729993e
Signed-off-by: Jean Christophe Beyler <jean.christophe.beyler@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/LoopInformation.cpp |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/vm/compiler/LoopInformation.cpp b/vm/compiler/LoopInformation.cpp
index 4f0c65b..3a065a9 100644
--- a/vm/compiler/LoopInformation.cpp
+++ b/vm/compiler/LoopInformation.cpp
@@ -241,6 +241,12 @@ void LoopInformation::fillBasicBlockVector (BasicBlock *current)
     //Mark it as visited
     dvmSetBit (basicBlocks, current->id);
 
+    //If it's a prebackward block, it's not in the loop
+    if (current->blockType == kPreBackwardBlock)
+    {
+        isInLoop = false;
+    }
+
     //If we are going back to first, we should be backward
     //But, we can also be a top loop with multiple backward branches
     //coming into us, and we are not done processing all our
-- 
1.7.4.1

